### ISSUE TITLE
Alignment

### ISSUE THREAD

**Issue by jhlusko** [Sleepwalking/pytorch-softdtw#8]
I would like to use softdtw for alignment, not as a loss function.

You mention "the derivative is the same as the expected DTW path".

I am new to pytorch, how do I get this derivative? I tried the following,

```
out = torch.tensor([0.0,0,1,0], requires_grad=True).reshape((1,4,1))
target = torch.tensor([0.0,1,0,0]).reshape((1,4,1))

criterion = SoftDTW(gamma=1.0)
loss = criterion(out, target)
m = loss.backward()
print(m)
```
which prints "None".

**Comment by jhlusko** [comment]
I have now altered _SoftDTW to the following,

```
class _SoftDTW(Function):
  @staticmethod
    def forward(ctx, D, gamma):
    dev = D.device
    dtype = D.dtype
    gamma = torch.Tensor([gamma]).to(dev).type(dtype) # dtype fixed
    D_ = D.detach().cpu().numpy()
    g_ = gamma.item()
    R = torch.Tensor(compute_softdtw(D_, g_)).to(dev).type(dtype)
    ctx.save_for_backward(D, R, gamma)
    # return R[:, -2, -2]
    # new from here
    D_ = D.detach().cpu().numpy()
    R_ = R.detach().cpu().numpy()
    g_ = gamma.item()
    E = torch.Tensor(compute_softdtw_backward(D, R, g_)).to(dev).type(dtype)
    return E
```
and removed the @jit(nopython = True) decorator from compute_softdtw_backward(D_, R, gamma).

This lets me compute the warping path to use in calculating my own loss function.

Since I changed _SoftDTW's forward method, do I need to change the backward method? What does the current return line do?

```
return grad_output.view(-1, 1, 1).expand_as(E) * E, None
```

Thanks for any help with this, I know its outside the scope of this project.

**Comment by Sleepwalking** [comment]
Hello @jhlusko. I actually have some unpushed changes pretty much the same as what you've done. In my case I just want a soft alignment matrix to weight the sequences without needing to pass the gradients through SoftDTW itself. The modification is trivial. The only difference is not modifying `forward` because that'll break the original SoftDTW. Instead, put these in a new method.
```
    @staticmethod
    def compute_expectation(D, gamma, bandwidth):
        dev = D.device
        dtype = D.dtype
        gamma = torch.Tensor([gamma]).to(dev).type(dtype)  # dtype fixed
        bandwidth = torch.Tensor([bandwidth]).to(dev).type(dtype)
        D_ = D.detach().cpu().numpy()
        g_ = gamma.item()
        b_ = bandwidth.item()
        R = torch.Tensor(compute_softdtw(D_, g_, b_)).to(dev).type(dtype)

        R_ = R.detach().cpu().numpy()
        E = torch.Tensor(compute_softdtw_backward(D_, R_, g_, b_)).to(dev).type(dtype)
        return E
```

However if you want to incorporate the expectation matrix into another loss function, that'll involve passing the gradient through the expectation. A easy solution is to remove the `detach().cpu().numpy()` calls and `@jit` decorators but doing so will make it painfully slow. I'm afraid there's no trivial solution.

**Comment by jhlusko** [comment]
Thanks for your response!

If I use @Maghoumi's pytorch-softdtw-cuda's implementation then it is fast, but limited to sequences < 1024 which I can live with.

Is it fine for gradient calculation to just put this in a new compute_expectation method? Don't I need to have that method extend torch.autograd.functional class with a forward and backward method? If so, any ideas what I need for the backward method?

**Comment by Sleepwalking** [comment]
> Is it fine for gradient calculation to just put this in a new compute_expectation method?

Let me break this down. For the original SoftDTW, the reason we don't just implement it in a single method, but choose to separately implement forward and backward method is for speed, not that we can't do it. There's a [nested loop](https://github.com/Sleepwalking/pytorch-softdtw/blob/master/soft_dtw.py#L13) that will cripple the performance due to overheads caused by a large number of PyTorch calls and Python interpreter itself.
To improve performance, I used Numba to JIT-compile that loop. There's a problem - Numba only works with Numpy arrays, not PyTorch tensors. So I had to wrap the optimized routines with Numpy-PyTorch type conversions. During the conversion, gradients are lost. That's why we had to separately implement a backward method to manually calculate the gradients.

Now, if you want a differentiable expectation implemented in a new method, you'll need to either (1) make the whole operation run in PyTorch (which means removing `detach().cpu().numpy()` and `@jit`), or (2) rewrite the backward calls.

Case (1) - with the entire operation done on PyTorch tensors, PyTorch will be able to auto differentiate `compute_expectation`. However, it will be very slow because of the unoptimized nested loop. Note that in this case the CUDA version won't work because the CUDA version also relies on JIT.

Case (2) - you'll need to derive `dE / dD` and update the backward method. This is like taking the hessian of the original SoftDTW.
